### PR TITLE
Add coreweave object store support.

### PR DESCRIPTION
**Pull request by eracah**
# What does this PR do?

Coreweave object store is compatible with boto3. Uploading objects to Coreweave object store is almost exactly like writing to S3 using S3Object Store, except an endpoint_url must be set: `https://object.<region>.coreweave.com`.

Unfortunately, the endpoint_url can only be set programmatically (AWS doesn't support an endpoint_url env var or config field), so we create our own env var called `AWS_ENDPOINT_URL`, that we check for in the `S3ObjectStore`. The Trainer interface for the user is exactly the same as normal S3, except they have to set this env var.

Tested manually for save and load checkpoint

# What issue(s) does this change relate to?

fix CO-1471

### CODE REVIEW

File: composer/utils/object_store/s3_object_store.py [hunk 1]
```diff
@@ -93,6 +93,9 @@ def __init__(
         if client_config is None:
             client_config = {}
         config = Config(**client_config)
+        if 'AWS_ENDPOINT_URL' in os.environ and endpoint_url is None:
```
Comment (dakinggg) [inline composer/utils/object_store/s3_object_store.py:96]: is `AWS_ENDPOINT_URL` the right name? Or is `S3_ENDPOINT_URL` more accurate?

**Review by dakinggg — APPROVED** [review]
LGTM assuming you did a manual test

File: docs/source/trainer/checkpointing.rst [hunk 2]
```diff
@@ -329,6 +329,34 @@ and delete the checkpoints from the local disk. The checkpoints will be located
 ``checkpoints/ep3.pt`` for third epoch's checkpoints, for example. The full URI in this case would be:
 ``s3://my_bucket/checkpoints/ep3.pt``.
 
+For uploading checkpoints to Coreweave's object store, the code is very similar to the
```
Comment (hanlint) [inline docs/source/trainer/checkpointing.rst:332]: can you link to some docs about Coreweave's object store?

File: docs/source/trainer/checkpointing.rst [hunk 3]
```diff
@@ -329,6 +329,34 @@ and delete the checkpoints from the local disk. The checkpoints will be located
...
+
+    os.environ['AWS_ENDPOINT_URL'] = 'https://object.las1.coreweave.com'
+    from composer.trainer import Trainer
+
+    # Save checkpoints every epoch to s3://my_bucket/checkpoints
+    trainer = Trainer(
+        model=model,
+        train_dataloader=train_dataloader,
+        max_duration='10ep',
+        save_folder='s3://my_bucket/checkpoints',
```
Comment (hanlint) [inline docs/source/trainer/checkpointing.rst:350]: Would it make more sense for the user to set a URI for coreweave (not sure what they use, we can ask them) That we use boto3/AWS to access coreweave seems like an implementation detail that the user should not need to be troubled with?

**Comment by mvpatel2000** [comment]
Is there a way to test this in a unit test? Do we want to start adding test coverage for things like this (non-S3 integration tests)? 

Not blocking, fine w merging, just food for thought

File: docs/source/trainer/checkpointing.rst (hunk 3 above)
Comment (eracah) [reply docs/source/trainer/checkpointing.rst:350]: yeah that's the weird thing is that they still use an s3 uri. See [here](https://docs.coreweave.com/storage/object-storage)

File: composer/utils/object_store/s3_object_store.py (hunk 1 above)
Comment (eracah) [reply composer/utils/object_store/s3_object_store.py]: oh good call. updated!

File: docs/source/trainer/checkpointing.rst (hunk 2 above)
Comment (eracah) [reply docs/source/trainer/checkpointing.rst]: done

**Comment by eracah** [comment]
> LGTM assuming you did a manual test

yeah did manual test of saving and loading checkpoints

**Comment by eracah** [comment]
> Is there a way to test this in a unit test? Do we want to start adding test coverage for things like this (non-S3 integration tests)?
> 
> Not blocking, fine w merging, just food for thought

Hmmm. Don't really think so because it is exactly like an S3 unit test, but the user sets the env variable, so not sure how we would make a new unit test

File: composer/utils/object_store/s3_object_store.py (hunk 1 above)
Comment (eracah) [reply composer/utils/object_store/s3_object_store.py]: also fixed in the mcli PR

File: docs/source/trainer/checkpointing.rst [hunk 4]
```diff
@@ -329,6 +329,34 @@ and delete the checkpoints from the local disk. The checkpoints will be located
 ``checkpoints/ep3.pt`` for third epoch's checkpoints, for example. The full URI in this case would be:
 ``s3://my_bucket/checkpoints/ep3.pt``.
 
+For uploading checkpoints to [Coreweave's object store](https://docs.coreweave.com/storage/object-storage), the code is very similar to the
+above S3 uploading code. The only difference is you must set your Coreweave endpoint url.
+To do this you can just set the ``AWS_ENDPOINT_URL`` environment variable before creating the
```
Comment (dakinggg) [inline docs/source/trainer/checkpointing.rst:334]: env var needs to be changed here too

Comment (eracah) [reply docs/source/trainer/checkpointing.rst]: oh right good catch

**Pull request merged**